/*
 * You may redistribute this program and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

syntax = "proto2";

package p2pnet.protocol;

message Handshake {
	enum Stage {
		INIT = 0; REPLY = 1;

		PUBKEY_ROTATION = 8;
		CONNECTION_REQUEST = 9;
	}
	required Stage stage = 1;

	message SignedHandshake {
		message KeyInfo {
			required bytes ecdsa_key = 2;

			required int64 expiration_time = 3;
			required int64 lose_time = 4;
		}
		required KeyInfo key_info = 1;
		optional bytes ecdh_key = 2;	// Not used in PUBKEY_ROTATION.
	}

	required SignedHandshake signed_handshake = 2;
	required bytes new_key_signature = 3;
	optional bytes old_key_signature = 4;	// For PUBKEY_ROTATION only.
}